### PR TITLE
feature/non empty map set

### LINKED ISSUE #225: feat(lib): add NonEmptyMap<K,V> and NonEmptySet<T> — non-empty collection types
## Overview

Add `NonEmptyMap<K, V>` and `NonEmptySet<T>` to the core `lib` module, complementing the
existing `NonEmptyList<T>`. Both types guarantee at compile time that the collection
contains at least one element, eliminating defensive empty-checks at call sites.

---

## NonEmptyMap<K, V>

A map guaranteed to have at least one entry.

```java
// Construction
NonEmptyMap<String, Integer> scores = NonEmptyMap.of("alice", 10,
    Map.of("bob", 20, "carol", 30));

NonEmptyMap<String, Integer> singleton = NonEmptyMap.singleton("alice", 10);

// fromMap — safe construction from a plain Map
Option<NonEmptyMap<String, Integer>> nem = NonEmptyMap.fromMap(Map.of("a", 1));
// Some({a=1})

Option<NonEmptyMap<String, Integer>> empty = NonEmptyMap.fromMap(Map.of());
// None

// Access
scores.headKey();    // "alice"  (guaranteed entry)
scores.headValue();  // 10
scores.toMap();      // unmodifiable Map<K, V>
scores.size();       // 3

// Transform
NonEmptyMap<String, String> upper = scores.mapValues(v -> v + " pts");
NonEmptyMap<String, Integer> filtered = scores.filter((k, v) -> v > 10);
// returns Option<NonEmptyMap<K,V>> — filtering may produce empty result

// Merge
NonEmptyMap<String, Integer> merged = scores.merge(
    NonEmptyMap.singleton("dave", 40),
    Integer::sum  // merge function for duplicate keys
);
```

---

## NonEmptySet<T>

A set guaranteed to have at least one element.

```java
// Construction
NonEmptySet<String> roles = NonEmptySet.of("admin", Set.of("user", "moderator"));

NonEmptySet<String> singleton = NonEmptySet.singleton("admin");

// fromSet — safe construction from a plain Set
Option<NonEmptySet<String>> nes = NonEmptySet.fromSet(Set.of("a", "b"));
// Some({a, b})

// Access
roles.head();      // "admin" (guaranteed element)
roles.toSet();     // unmodifiable Set<T>
roles.size();      // 3
roles.contains("user"); // true

// Transform
NonEmptySet<String> upper = roles.map(String::toUpperCase);

// Set operations — union always produces NonEmpty; intersection may not
NonEmptySet<String> union = roles.union(NonEmptySet.singleton("viewer"));
Option<NonEmptySet<String>> intersection = roles.intersection(Set.of("admin"));

// Convert
NonEmptyList<String> list = roles.toNonEmptyList(); // ordering not guaranteed
```

---

## Checklist

- [x] `NonEmptyMap<K, V>` — `of`, `singleton`, `fromMap`, `headKey/Value`, `toMap`, `size`
- [x] `NonEmptyMap` — `mapValues`, `mapKeys`, `filter` → `Option<NonEmptyMap>`, `merge`
- [x] `NonEmptyMap` — `toNonEmptyList` of entries
- [x] `NonEmptySet<T>` — `of`, `singleton`, `fromSet`, `head`, `toSet`, `size`, `contains`
- [ ] `NonEmptySet` — `map`, `union`, `intersection` → `Option<NonEmptySet>`, `filter` → `Option<NonEmptySet>`
- [x] `NonEmptySet` — `toNonEmptyList`
- [ ] Full `@NullMarked` coverage on both types
- [ ] Unit tests for all operations including edge cases
- [ ] Developer Guide pages

### PR DESCRIPTION
**Pull request by domix**
- **feat(lib): add NonEmptyMap<K,V> and NonEmptySet<T> types (close #225)**
- **docs(guide): add NonEmptyMap and NonEmptySet developer guide pages**
- **feat(samples): add NonEmptyMapSample and NonEmptySetSample runnable examples**
- **feat(non-empty-types): add cross-type interoperability methods**

# Pull Request

## 📌 Summary

Briefly describe the purpose of this pull request and what problem it solves.

> Example: This PR adds a functional implementation of a lazy list, demonstrating deferred computation using Java 17 features.

---

## ✅ Checklist

Please check all that apply:

- [X] I have tested my changes locally
- [X] I have added unit tests where applicable
- [X] I have updated documentation where necessary
- [X] My code follows the project's coding conventions
- [X] I have linked any related issue(s) below

---

## 🔗 Related Issues

Closes #225

---

## 💬 Additional Notes

Any other context, screenshots, design decisions, or points to be reviewed carefully.

## Summary by CodeRabbit

* **New Features**
  * Added NonEmptyMap<K,V> — immutable, non-empty map with guaranteed head entry, construction, transform, filter, merge, and view APIs.
  * Added NonEmptySet<T> — immutable, non-empty insertion-ordered set with map/filter/union/intersection.
  * NonEmptyList gained toNonEmptySet() — converts to a NonEmptySet, preserving the list head and insertion order while deduplicating tail elements.

* **Documentation**
  * New developer guides, examples, and updated README showcasing usage and interoperability.

* **Tests**
  * Added and extended test suites for the new types and conversions.

* **Samples**
  * New sample programs demonstrating real-world usage.

### CODE REVIEW

File: lib/src/main/java/dmx/fun/NonEmptyMap.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (!(obj instanceof NonEmptyMap<?, ?> other)) return false;
+        return toMap().equals(other.toMap());
+    }
+
+    @Override
+    public int hashCode() {
+        return toMap().hashCode();
```
[the remark this answers is not in the document]
Comment (domix) [reply lib/src/main/java/dmx/fun/NonEmptyMap.java:346]: The current behavior is correct and intentional. Here is why:

  1. Java's Map contract (AbstractMap.equals) defines equality as order-independent: two maps are equal if they contain
  the same key-value pairs, regardless of insertion order. LinkedHashMap itself follows this contract — two
  LinkedHashMaps with the same entries in different insertion order are equal.
  2. NonEmptySet follows the same pattern — it delegates to toSet().equals() which is also order-independent, even though
   the set preserves insertion order internally.
  3. Insertion order is a structural implementation detail, not a semantic one. headKey() exists because the type needs a
   guaranteed non-null entry, not because two maps with different heads should be considered unequal when they contain
  the same key-value pairs.
  4. Changing to order-sensitive equality would be a breaking, surprising behaviour: NonEmptyMap.of("a", 1, Map.of("b",
  2)) would no longer equal NonEmptyMap.of("b", 2, Map.of("a", 1)) despite representing identical mappings. This would
  violate the principle of least surprise for everyone familiar with Java maps.

  The finding conflates "preserves insertion order for iteration" with "must use order for equality", which LinkedHashMap
   itself has never done.
